src/ripple/core/impl/ThreadEntry.cpp

```diff
@@ -0,0 +1,66 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/core/ThreadEntry.h>
+
+#include <boost/coroutine/exceptions.hpp>
+
+namespace ripple {
+
+#ifndef NO_LOG_UNHANDLED_EXCEPTIONS
+namespace detail {
+thread_local
+std::string threadName;
+
+void setThreadName(std::string name)
+{
+    threadName = std::move(name);
+}
+}
+
+void terminateHandler()
+{
+    if (std::current_exception())
+    {
+        try
+        {
+            throw;
```

Comment:
Now that I'm thinking harder about this, it's really interesting that we can do a `throw/catch` here in the `terminateHandler` without losing the stack trace.  There's clearly some magic going on.  It's probably related to the need to call the `terminateHandler` no matter the reason for exiting (e.g., even for an uncaught exception).